src/include/catalog/catalog.h

```diff
@@ -70,85 +71,101 @@ class Catalog {
   // Global Singleton
   static Catalog *GetInstance(void);
 
-  Catalog();
-
-  // Creates the catalog database
-  void CreateCatalogDatabase(void);
-
-  // Creates the metric tables for statistics
-  void CreateMetricsCatalog();
+  // Deconstruct the catalog database when destroying the catalog.
+  ~Catalog();
 
   // Create a database
-  ResultType CreateDatabase(std::string database_name,
-                        concurrency::Transaction *txn);
+  ResultType CreateDatabase(const std::string &database_name,
+                            concurrency::Transaction *txn);
 
   // Add a database
   void AddDatabase(storage::Database *database);
 
-  // Add a database with name
-  void AddDatabase(std::string database_name,
-                   storage::Database *database);
-
   // Create a table in a database
-  ResultType CreateTable(std::string database_name, std::string table_name,
-                     std::unique_ptr<catalog::Schema>,
-                     concurrency::Transaction *txn);
+  ResultType CreateTable(const std::string &database_name,
+                         const std::string &table_name,
+                         std::unique_ptr<catalog::Schema>,
+                         concurrency::Transaction *txn);
 
   // Create the primary key index for a table
   ResultType CreatePrimaryIndex(const std::string &database_name,
-                            const std::string &table_name);
+                                const std::string &table_name,
+                                concurrency::Transaction *txn = nullptr);
+  ResultType CreatePrimaryIndex(oid_t database_oid, oid_t table_oid,
+                                concurrency::Transaction *txn = nullptr);
 
   ResultType CreateIndex(const std::string &database_name,
-                     const std::string &table_name,
-                     std::vector<std::string> index_attr,
-                     std::string index_name, bool unique, IndexType index_type);
+                         const std::string &table_name,
+                         const std::vector<std::string> &index_attr,
+                         const std::string &index_name, bool unique,
+                         IndexType index_type,
+                         concurrency::Transaction *txn = nullptr,
+                         bool is_catalog = false);
+  ResultType CreateIndex(oid_t database_oid, oid_t table_oid,
```

Comment:
Since this has a lot of input parameters, you should include comments that say what they all mean.